app/feature/feature-travel-certificate/src/main/kotlin/com/hedvig/android/feature/travelcertificate/ui/choose/ChooseContractForCertificateViewModel.kt

```diff
@@ -0,0 +1,72 @@
+package com.hedvig.android.feature.travelcertificate.ui.choose
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import com.hedvig.android.data.travelcertificate.ContractEligibleWithAddress
+import com.hedvig.android.data.travelcertificate.GetEligibleContractsWithAddressUseCase
+import com.hedvig.android.logger.LogPriority
+import com.hedvig.android.logger.logcat
+import com.hedvig.android.molecule.android.MoleculeViewModel
+import com.hedvig.android.molecule.public.MoleculePresenter
+import com.hedvig.android.molecule.public.MoleculePresenterScope
+
+internal class ChooseContractForCertificateViewModel(
+  getEligibleContractsWithAddressUseCase: GetEligibleContractsWithAddressUseCase,
+) : MoleculeViewModel<ChooseContractEvent, ChooseContractUiState>(
+    initialState = ChooseContractUiState.Loading,
+    presenter = ChooseContractPresenter(getEligibleContractsWithAddressUseCase),
+  )
+
+internal class ChooseContractPresenter(
+  private val getEligibleContractsWithAddressUseCase: GetEligibleContractsWithAddressUseCase,
+) : MoleculePresenter<ChooseContractEvent, ChooseContractUiState> {
+  @Composable
+  override fun MoleculePresenterScope<ChooseContractEvent>.present(
+    lastState: ChooseContractUiState,
+  ): ChooseContractUiState {
+    var dataLoadIteration by remember { mutableIntStateOf(0) }
+    var currentState by remember {
+      mutableStateOf<ChooseContractUiState>(ChooseContractUiState.Loading)
+    }
+
+    CollectEvents { event ->
+      when (event) {
+        ChooseContractEvent.RetryLoadData -> {
+          currentState = ChooseContractUiState.Loading
+          dataLoadIteration++
+        }
+      }
+    }
+
+    LaunchedEffect(dataLoadIteration) {
+      getEligibleContractsWithAddressUseCase.invoke().fold(
```

Comment:
My hunch here tells me that doing the loadIteration ++ and setting the state to loading is doing a bit too much inside the event collection itself.
I was thinking we could have the event just increment the iteration counter, and then inside the LaunchedEffect we can set it to loading, something like this
```
        ChooseContractEvent.RetryLoadData -> {
          dataLoadIteration++
        }
      }
    }

    LaunchedEffect(dataLoadIteration) {    
      currentState = ChooseContractUiState.Loading
      getEligibleContractsWithAddressUseCase.invoke().fold(
```

But then it would be true that if you for example come *back* into this screen, where your `lastState` would actually be `ChooseContractUiState.Success` already, then it would be jarring to have the state go into Loading mode, and then go back to the success as it was already, after the network request is finished.

So in the scenario of "I was one destination forward, then I came back into this screen and I want to just see what I saw before without any flickering into loading state and back into the success state", we got two options. We
1) Let the data fetching run again in this case, and let that happen in the background while we do not show that we are loading something to the user. If this succeeds, we get the success state back and we send it to the UI, but since it will be equivalent anyway most likely, the user will perceive nothing, all good. If the network request fails, then we do show the error state actually, despite not having shown that we are loading anything in the first place.
This could be done by doing this instead:
```
        ChooseContractEvent.RetryLoadData -> {
          dataLoadIteration++
        }
      }
    }

    LaunchedEffect(dataLoadIteration) {    
      if (currentState !is ChooseContractUiState.Success) {
        currentState = ChooseContractUiState.Loading
      }
      getEligibleContractsWithAddressUseCase.invoke().fold(
```

2) Could just not fetch any new data, aka never call `getEligibleContractsWithAddressUseCase.invoke()` if we just came back and we already have the success state available to us. Since to be honest the response from `getEligibleContractsWithAddressUseCase` should not have changed after going forward and back into this screen realistically.
This could be done like this:

```suggestion
        ChooseContractEvent.RetryLoadData -> {
          dataLoadIteration++
        }
      }
    }

    LaunchedEffect(dataLoadIteration) {
      if (currentState is ChooseContractUiState.Success) {
        return@LaunchedEffect
      }
      getEligibleContractsWithAddressUseCase.invoke().fold(
```
Note that here there is no way to fetch again even if the retry event comes in (it should never happen, since the retry button won't show in this scenario) if we are already in the `Success` state.





What do you think about all this? Do you prefer 1, 2, or something else?
Btw the current behavior with how you've set it up right now is more or less the same as 1), since coming back you will not set the state to loading, since that's only set in the event itself.

Reply:
I guess I like more number 1 in general, but, as you said, it is a kind of response that is not likely to be changed at all, so I think we better go with n.2 :)